hazelcast/src/main/java/com/hazelcast/cp/internal/MetadataRaftGroupSnapshot.java

```diff
@@ -27,9 +27,15 @@
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.unmodifiableCollection;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableSet;
 
 /**
  * Snapshot of the Metadata Raft group state
+ * This class should be IMMUTABLE since it is stored in the RaftLog's SnapshotEntry
```

Comment:
what about making the class truly immutable? Set everything in `final` fields in the constructor, no setters at all.

Reply:
This class is serialisable, so we should rewrite their de-serialisation.
The current issue comes from the getter.

I'll take a look if it is possible to use only the constructor and deep copy the parameters.

Reply:
ah right, 🤦 Let's just get rid of public setters to avoid accidental usage and use a constructor with all values as arguments in `MetadataRaftGroupManager#takeSnapshot`(like we do for other IDS classes - init via constructors with arguments at the point where it is created, use no-arg constructor just in IDS factory).